4/m_ipinfo_io.cpp

```diff
@@ -0,0 +1,189 @@
+            return;
+        }
+
+        std::string city = document.HasMember("city") ? document["city"].GetString() : "Unknown";
+        std::string region = document.HasMember("region") ? document["region"].GetString() : "Unknown";
+        std::string country = document.HasMember("country") ? document["country"].GetString() : "Unknown";
+        std::string org = document.HasMember("org") ? document["org"].GetString() : "Unknown";
+
+        std::string info = "City: " + city + ", Region: " + region + ", Country: " + country + ", Org: " + org;
+        cachedinfo.Set(user, info);
+
+        std::lock_guard<std::mutex> lock(mtx);
+        user->WriteNumeric(RPL_WHOISIPINFO, user->nick, "hes/her ip information: " + info);
+    }
+
+public:
+    IPInfoResolver(Module* Creator, LocalUser* user, StringExtItem& cache, const std::string& key)
+        : cachedinfo(cache), apikey(key), theiruuid(user->uuid)
+    {
+        worker = std::thread(&IPInfoResolver::DoRequest, this, user);
+    }
+
+    ~IPInfoResolver()
+    {
+        if (worker.joinable())
+        {
+            worker.join();
+        }
+    }
+};
+
+class ModuleIPInfo : public Module, public Stats::EventListener, public Whois::EventListener
+{
+private:
+    StringExtItem cachedinfo;
+    std::string apikey;
+
+public:
+    ModuleIPInfo()
+        : Module(VF_VENDOR, "Adds IPinfo.io information to WHOIS responses for opers, using a configured API key.")
+        , Stats::EventListener(this)
+        , Whois::EventListener(this)
+        , cachedinfo(this, "ipinfo", ExtensionType::USER)
+    {
+    }
+
+    void ReadConfig(ConfigStatus& status) override
+    {
+        auto& tag = ServerInstance->Config->ConfValue("ipinfo");
+        apikey = tag->getString("apikey", "", 1);
+
+        const UserManager::LocalList& users = ServerInstance->Users.GetLocalUsers();
+        for (const auto& user : users)
+        {
+            cachedinfo.Unset(user);
+        }
+    }
+
+    void OnWhois(Whois::Context& whois) override
+    {
+        User* target = whois.GetTarget();
+
+        if (target->server->IsService() || target->IsModeSet('B'))
+            return;
+
+        if (!whois.GetSource()->IsOper())
+            return;
+
+        const std::string* cached = cachedinfo.Get(target);
+        if (cached)
+        {
+            whois.SendLine(RPL_WHOISIPINFO, "hes/her ip information(cached): " + *cached);
+        }
+        else
+        {
+            new IPInfoResolver(this, IS_LOCAL(target), cachedinfo, apikey);
+        }
+    }
+
+    ModResult OnStats(Stats::Context& stats) override
+    {
+        return MOD_RES_PASSTHRU;
+    }
+
+    void OnUnloadModule(Module* mod) override
+    {
+        const UserManager::LocalList& users = ServerInstance->Users.GetLocalUsers();
+        for (const auto& user : users)
+        {
+            cachedinfo.Unset(user);
+        }
+    }
```

Comment:
Extensions are automatically unset when the providing module is unloaded. This is unnecessary.